app/src/main/java/crux/bphc/cms/fragments/MyCoursesFragment.kt

```diff
@@ -0,0 +1,418 @@
+            try {
+                val courseList = courseRequestHandler.fetchCourseListSync()
+                courses.clear()
+                courses.addAll(courseList)
+                val realm = Realm.getDefaultInstance() // tie a realm instance to this thread
+                val courseDataHandler = CourseDataHandler(requireContext(), realm)
+                courseDataHandler.replaceCourses(courseList)
+                realm.close()
+                checkEmpty()
+                updateCourseContent()
+            } catch (e : Exception) {
+                Log.e(TAG, "", e)
+                CoroutineScope(Dispatchers.Main).launch {
+                    Toast.makeText(requireActivity(), "Error: ${e.message}", Toast.LENGTH_SHORT).show()
+                    if (e is InvalidTokenException) {
+                        UserUtils.logout(requireActivity())
+                        UserUtils.clearBackStackAndLaunchTokenActivity(requireActivity())
+                    }
+                }
+            } finally {
+                CoroutineScope(Dispatchers.Main).launch {
+                    swipeRefreshLayout.isRefreshing = false
+                }
+            }
+        }
+    }
+
+    private suspend fun updateCourseContent() {
+        coroutineScope {
+            val courseRequestHandler = CourseRequestHandler(requireActivity())
+            val promises = courses.map map@ {
+                async innerAsync@{
+                    val sections: MutableList<CourseSection>
+                    try {
+                        sections = courseRequestHandler.getCourseDataSync(it.courseId)
+                    } catch (e: IOException) {
+                        Log.e(TAG, "IOException when syncing course: ${it.courseId}}", e)
+                        return@innerAsync false
+                    }
+
+                    val realm = Realm.getDefaultInstance() // tie a realm instance to this thread
+                    val courseDataHandler = CourseDataHandler(requireContext(), realm)
+
+                    for (courseSection in sections) {
+                        val modules = courseSection.modules
+                        for (module in modules) {
+                            if (module.modType == Module.Type.FORUM) {
+                                val discussions = courseRequestHandler
+                                        .getForumDicussionsSync(module.instance)
+                                for (d in discussions) {
+                                    d.setForumId(module.instance)
+                                }
+                            }
+                        }
+                    }
+
+                    val newPartsInSections = courseDataHandler
+                            .isolateNewCourseData(it.courseId, sections)
+                    courseDataHandler.replaceCourseData(it.courseId, sections)
+
+                    realm.close() // let's not forget to do this
+                    if (newPartsInSections.isNotEmpty()) {
```

Comment:
You can return a 1 for true and 0 for false here and then later on while "folding" it, directly use the values. It's a minor thing though.

Reply:
That is what I did initially. But that made less sense than using `true` or `false` to indicate that a course has updates.

Reply:
True, it isn't good readability wise